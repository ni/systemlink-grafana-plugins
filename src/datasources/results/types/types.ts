import { DataQuery, DataSourceJsonData } from '@grafana/schema';

export interface ResultsQuery extends DataQuery {
  queryType: QueryType;
}

export enum QueryType {
  Results = 'Results',
  Steps = 'Steps'
}

export enum OutputType {
  Data = 'Data',
  TotalCount = 'Total Count'
}

export enum UseTimeRangeFor {
  Started = 'Started',
  Updated = 'Updated'
}

export interface ResultsFeatureToggles {
  queryByResults: boolean;
  queryBySteps: boolean;
}

export interface ResultsDataSourceOptions extends DataSourceJsonData {
  featureToggles: ResultsFeatureToggles;
}

export const ResultsFeatureTogglesDefaults: ResultsFeatureToggles = {
  queryByResults: true,
  queryBySteps: true
}
