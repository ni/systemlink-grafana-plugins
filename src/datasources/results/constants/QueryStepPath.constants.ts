export const QUERY_PATH_REQUEST_PER_SECOND = 6;
export const MAX_PATH_TAKE_PER_REQUEST = 1000;
