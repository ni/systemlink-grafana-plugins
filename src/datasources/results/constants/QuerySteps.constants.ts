export const QUERY_STEPS_REQUEST_PER_SECOND = 6;
export const MAX_TAKE_PER_REQUEST = 500;
