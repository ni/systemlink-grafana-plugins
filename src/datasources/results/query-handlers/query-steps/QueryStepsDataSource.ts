import { DataQueryRequest, DataFrameDTO, FieldType } from '@grafana/data';
import { OutputType } from 'datasources/results/types/types';
import {
  QueryResponse,
  QuerySteps,
  QueryStepsResponse,
  StepsProperties,
  StepsPropertiesOptions,
  StepsResponseProperties,
} from 'datasources/results/types/QuerySteps.types';
import { ResultsDataSourceBase } from 'datasources/results/ResultsDataSourceBase';
import { defaultStepsQuery } from 'datasources/results/defaultQueries';
import { MAX_TAKE_PER_REQUEST, QUERY_STEPS_REQUEST_PER_SECOND } from 'datasources/results/constants/QuerySteps.constants';

export class QueryStepsDataSource extends ResultsDataSourceBase {
  queryStepsUrl = this.baseUrl + '/v2/query-steps';

  defaultQuery = defaultStepsQuery;

  async querySteps(
    filter?: string,
    orderBy?: string,
    projection?: StepsProperties[],
    take?: number,
    descending?: boolean,
    continuationToken?: string,
    returnCount = false
  ): Promise<QueryStepsResponse> {

    const response = await this.post<QueryStepsResponse>(`${this.queryStepsUrl}`, {
      filter,
      orderBy,
      descending,
      projection,
      take,
      continuationToken,
      returnCount,
    });

    if (response.error) {
      throw new Error(`An error occurred while querying steps: ${response.error}`);
    }

    return response;
  }

  async queryStepsInBatches(
    filter?: string,
    orderBy?: string,
    projection?: StepsProperties[],
    take?: number,
    descending?: boolean,
    returnCount = false,
  ): Promise<QueryStepsResponse> {
    const queryRecord = async (currentTake: number, token?: string): Promise<QueryResponse<StepsResponseProperties>> => {
      const response = await this.querySteps(
        filter,
        orderBy,
        projection,
        currentTake,
        descending,
        token,
        returnCount
      );

      return {
        data: response.steps,
        continuationToken: response.continuationToken,
        totalCount: response.totalCount
      };
    };

    const batchQueryConfig = {
      maxTakePerRequest: MAX_TAKE_PER_REQUEST,
      requestsPerSecond: QUERY_STEPS_REQUEST_PER_SECOND
    };

    const response = await this.queryInBatches(queryRecord, batchQueryConfig, take);

    return {
      steps: response.data,
      continuationToken: response.continuationToken,
      totalCount: response.totalCount
    };
  }
  
  async runQuery(query: QuerySteps, options: DataQueryRequest): Promise<DataFrameDTO> {
    const projection = query.showMeasurements
      ? [...new Set([...(query.properties || []), StepsPropertiesOptions.DATA])]
      : query.properties;

    if (query.outputType === OutputType.Data) {
      const responseData = await this.queryStepsInBatches(
        this.getTimeRangeFilter(options, query.useTimeRange, query.useTimeRangeFor),
        query.orderBy,
        projection as StepsProperties[],
        query.recordCount,
        query.descending,
        true
      );
  
      if (responseData.steps.length === 0) {
        return {
          refId: query.refId,
          fields: [],
        };
      }
      const stepsResponse = responseData.steps;
      const stepResponseKeys = new Set(Object.keys(stepsResponse[0]));
      const selectedFields = (query.properties || []).filter(field => stepResponseKeys.has(field));
      const fields = this.processFields(selectedFields, stepsResponse);

      if (query.showMeasurements) {
        const measurementFields = this.processMeasurementData(stepsResponse);
        fields.push(...measurementFields);
      }
      return {
        refId: query.refId,
        fields: fields,
      };
    } else {
      const responseData = await this.querySteps(
        this.getTimeRangeFilter(options, query.useTimeRange, query.useTimeRangeFor),
        undefined,
        undefined,
        undefined,
        undefined,
        undefined,
        true
      );

      return {
        refId: query.refId,
        fields: [{ name: 'Total count', values: [responseData.totalCount] }],
      };
    }
  }

  private processFields(
    selectedFields: StepsProperties[],
    stepsResponse: StepsResponseProperties[]
  ): Array<{ name: StepsProperties; values: string[]; type: FieldType }> {
    return selectedFields.map(field => {
      const isTimeField = field === StepsPropertiesOptions.UPDATED_AT || field === StepsPropertiesOptions.STARTED_AT;

      const fieldType = isTimeField ? FieldType.time : FieldType.string;
      const values = stepsResponse.map(data => data[field as keyof StepsResponseProperties]);

      switch (field) {
        case StepsPropertiesOptions.PROPERTIES:
        case StepsPropertiesOptions.INPUTS:
        case StepsPropertiesOptions.OUTPUTS:
        case StepsPropertiesOptions.DATA:
          return {
            name: field,
            values: values.map(value => (value !== null ? JSON.stringify(value) : '')),
            type: fieldType,
          };
        case StepsPropertiesOptions.STATUS:
          return { name: field, values: values.map((value: any) => value?.statusType), type: fieldType };
        default:
          return { name: field, values, type: fieldType };
      }
    });
  }

  private processMeasurementData(stepsResponse: StepsResponseProperties[]): any[] {
    const fieldToParameterProperty = {
      'Measurement Name': 'name',
      'Measurement Value': 'measurement',
      'Status': 'status',
      'Unit': 'units',
      'Low Limit': 'lowLimit',
      'High Limit': 'highLimit',
    };
    const measurementFields = Object.keys(fieldToParameterProperty) as Array<keyof typeof fieldToParameterProperty>;

    return measurementFields.map(measurementField => {
      const values = stepsResponse.map(step => {
        if (!step.data?.parameters) {
          return [];
        }
        return step.data.parameters.map(
          param => param[fieldToParameterProperty[measurementField]]
        );
      });

      return {
        name: measurementField,
        values,
        type: FieldType.string,
      };
    });
  }

  shouldRunQuery(_: QuerySteps): boolean {
    return true;
  }
}
