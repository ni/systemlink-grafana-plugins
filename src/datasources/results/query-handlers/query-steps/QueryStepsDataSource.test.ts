import { MockProxy } from 'jest-mock-extended';
import { BackendSrv, TemplateSrv } from '@grafana/runtime';
import { createFetchError, createFetchResponse, getQueryBuilder, requestMatching, setupDataSource } from 'test/fixtures';
import { Field } from '@grafana/data';
import { QuerySteps, QueryStepsResponse, StepsProperties, StepsPropertiesOptions } from 'datasources/results/types/QuerySteps.types';
import { OutputType, QueryType } from 'datasources/results/types/types';
import { QueryStepsDataSource } from './QueryStepsDataSource';
import { ResultsQueryBuilderFieldNames } from 'datasources/results/constants/ResultsQueryBuilder.constants';
import { StepsQueryBuilderFieldNames } from 'datasources/results/constants/StepsQueryBuilder.constants';
import { StepsVariableQuery } from 'datasources/results/types/QueryResults.types';

const mockQueryStepsResponse: QueryStepsResponse = {
  steps: [
    {
      stepId: '1',
      name: 'Step 1',
      properties: {
        key1: 'value1',
        key2: 'value2',
      },
    },
  ],
  continuationToken: undefined,
  totalCount: 1
};

let datastore: QueryStepsDataSource, backendServer: MockProxy<BackendSrv>, templateSrv: MockProxy<TemplateSrv>;

jest.mock('datasources/results/constants/QuerySteps.constants', () => ({
  ...jest.requireActual('datasources/results/constants/QuerySteps.constants'),
  QUERY_STEPS_REQUEST_PER_SECOND: 2,
}));

describe('QueryStepsDataSource', () => {
  beforeEach(() => {
    [datastore, backendServer, templateSrv] = setupDataSource(QueryStepsDataSource);

    backendServer.fetch
      .calledWith(requestMatching({ url: '/nitestmonitor/v2/query-steps', method: 'POST' }))
      .mockReturnValue(createFetchResponse(mockQueryStepsResponse));
  })

  afterEach(() => {
    jest.useRealTimers();
  });

  describe('querySteps', () => {
    test('should return data when there are valid queries', async () => {
      const response = await datastore.querySteps();

      expect(response).toMatchSnapshot();
    });

    test('should raise an error when API fails', async () => {
      backendServer.fetch
        .calledWith(requestMatching({ url: '/nitestmonitor/v2/query-steps' }))
        .mockReturnValue(createFetchError(400));

      await expect(datastore.querySteps())
        .rejects
        .toThrow('Request to url "/nitestmonitor/v2/query-steps" failed with status code: 400. Error message: "Error"');
    });
  });

  describe('query', () => {
    test('should return data for valid data-output-type query', async () => {
      const query = buildQuery({
        refId: 'A',
        outputType: OutputType.Data
      });

      const response = await datastore.query(query);

      expect(response.data).toHaveLength(1);
      expect(response.data).toMatchSnapshot();
    });

    test('should return total count for valid total count output type queries', async () => {
      const query = buildQuery({
        refId: 'A',
        outputType: OutputType.TotalCount,
      });

      const response = await datastore.query(query);

      expect(response.data).toHaveLength(1);
      expect(response.data).toMatchSnapshot();
    });

    test('should return no data when QuerySteps API returns empty array', async () => {
      backendServer.fetch
        .calledWith(requestMatching({ url: '/nitestmonitor/v2/query-steps', method: 'POST' }))
        .mockReturnValue(
          createFetchResponse(
            {
              steps: [],
              continuationToken: null,
              totalCount: 0
            } as unknown as QueryStepsResponse
          )
        );

      const query = buildQuery();
      const response = await datastore.query(query);

      expect(response.data).toMatchSnapshot();
    });

    test('should return no data when Query Steps returns error', async () => {
      backendServer.fetch
        .calledWith(requestMatching({ url: '/nitestmonitor/v2/query-steps' }))
        .mockReturnValue(createFetchError(400));

      const query = buildQuery(
        {
          refId: 'A',
          outputType: OutputType.Data
        },
      );

      await expect(datastore.query(query))
        .rejects
        .toThrow('Request to url "/nitestmonitor/v2/query-steps" failed with status code: 400. Error message: "Error"');
    });

    test('should convert properties to Grafana fields', async () => {
      const query = buildQuery(
        {
          refId: 'A',
          outputType: OutputType.Data,
          properties: [StepsPropertiesOptions.PROPERTIES as StepsProperties]
        },
      );

      const response = await datastore.query(query);

      const fields = response.data[0].fields as Field[];
      expect(fields).toMatchSnapshot();
    });

    test('should convert step measurements to Grafana fields when show measurments is enabled', async () => {
      const mockQueryStepsMeasurementResponse: QueryStepsResponse = {
        steps: [
          {
            stepId: '1',
            data: {
              text: 'Step 1',
              parameters: [
                {
                  name: 'Voltage',
                  measurement: '3.7',
                  status: 'Passed',
                  units: 'V',
                  lowLimit: '3.5',
                  highLimit: '4.0',
                  value: ''
                },
                {
                  name: 'Voltage',
                  measurement: '3.7',
                  status: 'Passed',
                  units: 'V',
                  lowLimit: '3.5',
                  highLimit: '4.0',
                  value: ''
                },//duplicate measurement
                {
                  name: 'Current',
                  measurement: '1.2',
                  status: 'Failed',
                  units: 'A',
                  lowLimit: '1.0',
                  highLimit: '1.5',
                  miscellaneous: 'Misc'
                },
              ]
            }
          },
        ],
        totalCount: 1
      };

      backendServer.fetch
        .calledWith(requestMatching({ url: '/nitestmonitor/v2/query-steps', method: 'POST' }))
        .mockReturnValue(createFetchResponse(mockQueryStepsMeasurementResponse));

      const query = buildQuery(
        {
          refId: 'A',
          outputType: OutputType.Data,
          showMeasurements: true
        },
      );

      const response = await datastore.query(query);

      const fields = response.data[0].fields as Field[];
      expect(fields).toMatchSnapshot();
    });

    test('should include templateSrv replaced values in the filter', async () => {
      const timeRange = {
        Started: 'startedAt',
        Updated: 'updatedAt',
      }
      const selectedUseTimeRangeFor = 'Started';
      const filter = `(${timeRange[selectedUseTimeRangeFor]} > "\${__from:date}" && ${timeRange[selectedUseTimeRangeFor]} < "\${__to:date}")`;
      const replacedFilter = `(${timeRange[selectedUseTimeRangeFor]} > "2025-04-01" && ${timeRange[selectedUseTimeRangeFor]} < "2025-04-02")`;
      templateSrv.replace.calledWith().mockReturnValue(replacedFilter);

      const query = buildQuery(
        {
          refId: 'A',
          outputType: OutputType.Data,
          useTimeRange: true,
          useTimeRangeFor: selectedUseTimeRangeFor
        },
      );

      await datastore.query(query);

      expect(templateSrv.replace).toHaveBeenCalledWith(filter, expect.anything());
      expect(backendServer.fetch).toHaveBeenCalledWith(
        expect.objectContaining({
          data: expect.objectContaining({ filter: replacedFilter }),
        })
      );
    });

    test('should handle null and undefined properties', async () => {
      backendServer.fetch
        .calledWith(requestMatching({ url: '/nitestmonitor/v2/query-steps' }))
        .mockReturnValue(createFetchResponse({
          steps: [
            {
              id: '1',
              properties: null
            }
          ], continuationToken: null, totalCount: 1
        } as unknown as QueryStepsResponse));

      const query = buildQuery(
        {
          refId: 'A',
          outputType: OutputType.Data,
          properties: [
            StepsPropertiesOptions.PROPERTIES
          ] as StepsProperties[],
          orderBy: undefined
        },
      );

      const response = await datastore.query(query);
      const fields = response.data[0].fields as Field[];
      expect(fields).toEqual([
        { name: 'properties', values: [""], type: 'string' },
      ]);
    });

    test('should call query steps API once when output type is total count', async () => {
      const mockResponses = [
        createFetchResponse({
          steps: Array(100).fill({ stepId: '1', name: 'Step 1' }),
          continuationToken: null,
          totalCount: 5000,
        }),
        createFetchResponse({
          steps: Array(100).fill({ stepId: '2', name: 'Step 2' }),
          continuationToken: null,
          totalCount: 5000,
        })
      ]
      backendServer.fetch
        .mockImplementationOnce(() => mockResponses[0])
        .mockImplementationOnce(() => mockResponses[1])
      const query = buildQuery(
        {
          refId: 'A',
          outputType: OutputType.TotalCount,
          properties: [
            StepsPropertiesOptions.PROPERTIES
          ] as StepsProperties[],
          orderBy: undefined,
          useTimeRange: true,
          useTimeRangeFor: 'Updated'
        },
      );

      const response = await datastore.query(query);

      const fields = response.data[0].fields as Field[];
      expect(fields).toEqual([
        { name: 'Total count', values: [5000] },
      ]);
      expect(backendServer.fetch).toHaveBeenCalledTimes(1);
    });
  });

  it('should not call query-steps when resultsQuery is empty',async () => {
    const query = buildQuery({
      refId: 'A',
      outputType: OutputType.Data,
      resultsQuery: '',
    });

    const response = await datastore.query(query);

    const fields = response.data[0].fields as Field[];
    expect(fields).toEqual([]);
    expect(backendServer.fetch).not.toHaveBeenCalled();
  });

  describe('fetch Steps with rate limiting', () => {
    it('should make a single request when take is less than MAX_TAKE_PER_REQUEST', async () => {
      const mockResponses = [
        createFetchResponse({
          steps: Array(100).fill({ stepId: '1', name: 'Step 1' }),
          continuationToken: null,
          totalCount: 100,
      })]
      backendServer.fetch
        .mockImplementationOnce(() => mockResponses[0])
        const responsePromise = datastore.queryStepsInBatches(
          undefined,
          undefined,
          undefined,
          100,
          undefined,
          undefined,
          true
        );
        const response = await responsePromise;
  
        expect(response.steps).toHaveLength(100);
        expect(backendServer.fetch).toHaveBeenCalledTimes(1);
        expect(backendServer.fetch).toHaveBeenNthCalledWith(
          1,
          expect.objectContaining({
            data: expect.objectContaining({ take: 100, continuationToken: undefined }),
          })
        );
      });

      test('should batch requests when total number od steps matching the filter is less than requested take', async () => {
        const mockResponses = [
          createFetchResponse({
            steps: Array(500).fill({ stepId: '1', name: 'Step 1' }),
            continuationToken: 'token1',
            totalCount: 900,
          }),
          createFetchResponse({
            steps: Array(400).fill({ stepId: '1', name: 'Step 1' }),
            continuationToken: 'token2',
            totalCount: 900,
          })
        ]
        backendServer.fetch
          .mockImplementationOnce(() => mockResponses[0])
          .mockImplementationOnce(() => mockResponses[1])
          const responsePromise = datastore.queryStepsInBatches(
            'name = \"test\"',
            undefined,
            undefined,
            10000,
            undefined,
            undefined,
            true
          );
          const response = await responsePromise;
    
          expect(response.steps).toHaveLength(900);
          expect(backendServer.fetch).toHaveBeenCalledTimes(2);
          expect(backendServer.fetch).toHaveBeenNthCalledWith(
            1,
            expect.objectContaining({
              data: expect.objectContaining({ take: 500, continuationToken: undefined }),
            })
          );
          expect(backendServer.fetch).toHaveBeenNthCalledWith(
            2,
            expect.objectContaining({
              data: expect.objectContaining({ take: 400, continuationToken: 'token1' }),
            })
          );
      })

      test('should batch requests with RequestPerSecond', async () => {
        jest.useFakeTimers();
        const fetchSpy = jest.spyOn(backendServer, 'fetch');
        
        const mockResponses = [
          createFetchResponse({
            steps: Array(500).fill({ stepId: '1', name: 'Step 1' }),
            continuationToken: 'token1',
            totalCount: 2000,
          }),
          createFetchResponse({
            steps: Array(500).fill({ stepId: '2', name: 'Step 2' }),
            continuationToken: 'token2', 
            totalCount: 2000,
          }),
          createFetchResponse({
            steps: Array(500).fill({ stepId: '3', name: 'Step 3' }),
            continuationToken: 'token3',
            totalCount: 2000,
          }),
          createFetchResponse({
            steps: Array(500).fill({ stepId: '4', name: 'Step 4' }),
            continuationToken: null,
            totalCount: 2000,
          })
        ];

        backendServer.fetch
          .mockImplementationOnce(() => mockResponses[0])
          .mockImplementationOnce(() => mockResponses[1])
          .mockImplementationOnce(() => mockResponses[2])
          .mockImplementationOnce(() => mockResponses[3])
          
        const responsePromise = datastore.queryStepsInBatches(
          undefined,
          undefined,
          undefined,
          2000,
          undefined,
          undefined,
          true
        );

        await jest.advanceTimersByTimeAsync(0);
        expect(fetchSpy).toHaveBeenCalledTimes(2);
        await jest.advanceTimersByTimeAsync(1000);
        expect(fetchSpy).toHaveBeenCalledTimes(4);
        await responsePromise;

        expect(fetchSpy).toHaveBeenNthCalledWith(
          1,
          expect.objectContaining({
            data: expect.objectContaining({ take: 500, continuationToken: undefined }),
          })
        );
        expect(fetchSpy).toHaveBeenNthCalledWith(
          2,
          expect.objectContaining({
            data: expect.objectContaining({ take: 500, continuationToken: 'token1' }),
          })
        );
        expect(fetchSpy).toHaveBeenNthCalledWith(
          3,
          expect.objectContaining({
            data: expect.objectContaining({ take: 500, continuationToken: 'token2' }),
          })
        );
        expect(fetchSpy).toHaveBeenNthCalledWith(
          4,
          expect.objectContaining({
            data: expect.objectContaining({ take: 500, continuationToken: 'token3' }),
          })
        );
      });
  
      test('should stop fetching when continuationToken is null', async () => {
        const mockResponses = [
          createFetchResponse({
            steps: Array(500).fill({ stepId: '1', name: 'Step 1' }),
            continuationToken: null,
            totalCount: 500,
          }),
        ];
  
        backendServer.fetch.mockImplementationOnce(() => mockResponses[0]);
        const response = await datastore.queryStepsInBatches(
          undefined,
          undefined,
          undefined,
          3000,
          undefined,
          undefined,
          true
        );
  
        expect(response.steps).toHaveLength(500);
        expect(backendServer.fetch).toHaveBeenCalledTimes(1);
        expect(backendServer.fetch).toHaveBeenCalledWith(
          expect.objectContaining({
            data: expect.objectContaining({ take: 500, continuationToken: undefined }),
          })
        );
      });

      test('should stop subsequent API calls after error occurs', async () => {
        backendServer.fetch
          .mockImplementationOnce(() => createFetchResponse({
            steps: Array(500).fill({ stepId: '1', name: 'Step 1' }),
            continuationToken: 'token1',
            totalCount: 1500,
          }))
          .mockImplementationOnce(() => createFetchError(400)) //Error
          .mockImplementationOnce(() => createFetchResponse({
            steps: Array(500).fill({ stepId: '2', name: 'Step 2' }),
            continuationToken: 'token2',
            totalCount: 1500,
          }));
        
        const batchPromise = datastore.queryStepsInBatches(
          'filter',
          'orderBy',
          undefined,
          1500,
          false,
          undefined,
          true
        );
        
        await expect(batchPromise)
          .rejects
          .toThrow('Request to url "/nitestmonitor/v2/query-steps" failed with status code: 400. Error message: "Error"');
        expect(backendServer.fetch).toHaveBeenCalledTimes(2);
      });

      test('should delay between consecutive batch API calls', async () => {
        const mockTimeValues = [1000, 1200];
        let timeCallCount = 0;
        
        Date.now = jest.fn().mockImplementation(() => {
          return mockTimeValues[timeCallCount++] || mockTimeValues[mockTimeValues.length - 1];
        });
        
        const spyDelay = jest.spyOn(datastore as any, 'delay');

        const mockResponses = [
          createFetchResponse({
            steps: Array(500).fill({ stepId: '1', name: 'Step 1' }),
            continuationToken: 'token1',
            totalCount: 2000,
          }),
          createFetchResponse({
            steps: Array(500).fill({ stepId: '2', name: 'Step 2' }),
            continuationToken: 'token2',
            totalCount: 2000,
          }),
          createFetchResponse({
            steps: Array(500).fill({ stepId: '3', name: 'Step 3' }),
            continuationToken: 'token3',
            totalCount: 2000,
          }),
          createFetchResponse({
            steps: Array(500).fill({ stepId: '4', name: 'Step 3' }),
            continuationToken: null,
            totalCount: 2000,
          }),
        ];
        backendServer.fetch
          .mockImplementationOnce(() => mockResponses[0])
          .mockImplementationOnce(() => mockResponses[1])
          .mockImplementationOnce(() => mockResponses[2])
          .mockImplementationOnce(() => mockResponses[3])
        const responsePromise = datastore.queryStepsInBatches(
          undefined,
          undefined,
          undefined,
          2000,
          undefined,
          undefined,
          true
        );

        const response = await responsePromise;
  
        expect(response.steps).toHaveLength(2000);
        expect(backendServer.fetch).toHaveBeenCalledTimes(4);
        expect(spyDelay).toHaveBeenCalledTimes(1);
        expect(spyDelay).toHaveBeenCalledWith(800); // delay for 1000 - 200 = 800ms
      });
    });

    describe('query builder queries', () => {
      test('should transform the resultsfilter and stepsfilter contains single query', async () => {
        const query = buildQuery({
          refId: 'A',
          outputType: OutputType.Data,
          resultsQuery: `${ResultsQueryBuilderFieldNames.PART_NUMBER} = "partNumber1"`,
          stepsQuery: `${StepsQueryBuilderFieldNames.TYPE} = "Type1"`
        })
        await datastore.query(query);

        expect(backendServer.fetch).toHaveBeenCalledWith(
          expect.objectContaining({
            url: '/nitestmonitor/v2/query-steps',
            data: expect.objectContaining({
              resultsFilter: "PartNumber = \"partNumber1\"",
              filter: "stepType = \"Type1\""
            }),
          })
        );
      });

      test('should transform fields when contains multiple queries', async () => {
        const query = buildQuery({
          refId: 'A',
          outputType: OutputType.Data,
          resultsQuery: `${ResultsQueryBuilderFieldNames.PART_NUMBER} = "{partNumber1,partNumber2}"`
        })
        await datastore.query(query);

        expect(backendServer.fetch).toHaveBeenCalledWith(
          expect.objectContaining({
            url: '/nitestmonitor/v2/query-steps',
            data: expect.objectContaining({
              resultsFilter: "(PartNumber = \"partNumber1\" || PartNumber = \"partNumber2\")",
            }),
          })
        );
      });

      test('should transform fields when queryBy contains a date', async () => {   
        jest.useFakeTimers().setSystemTime(new Date('2025-01-01'));     

        const query = buildQuery(
          {
            resultsQuery: 'UpdatedAt = "${__now:date}"',
            stepsQuery: 'StartedAt = "${__now:date}"',
          },
        );

        await datastore.query(query);

        expect(backendServer.fetch).toHaveBeenCalledWith(
          expect.objectContaining({
            url: '/nitestmonitor/v2/query-steps',
            data: expect.objectContaining({
              resultsFilter: 'UpdatedAt = "2025-01-01T00:00:00.000Z"'
            }),
          })
        );

        jest.useRealTimers();
      });

      test('should transform query when queryBy contains nested expressions', async () => {
        const query = buildQuery(
          {
            refId: 'A',
            resultsQuery: `(${ResultsQueryBuilderFieldNames.PART_NUMBER} = "123" || ${ResultsQueryBuilderFieldNames.KEYWORDS} != "456") && ${ResultsQueryBuilderFieldNames.HOSTNAME} contains "Test"`,
            stepsQuery: `(${StepsQueryBuilderFieldNames.TYPE} = "123" || ${StepsQueryBuilderFieldNames.KEYWORDS} != "456") && ${StepsQueryBuilderFieldNames.NAME} contains "Test"`
          },
        );

        await datastore.query(query);

        expect(backendServer.fetch).toHaveBeenCalledWith(
          expect.objectContaining({
            url: '/nitestmonitor/v2/query-steps',
            data: expect.objectContaining({
              resultsFilter:  "(PartNumber = \"123\" || Keywords != \"456\") && HostName contains \"Test\"",
              filter: "(stepType = \"123\" || keywords != \"456\") && name contains \"Test\""
            }),
          })
        );
      });
    });

    describe('metricFindQuery', () => {
      it('should return empty array if queryByResults and queryBySteps are undefined', async () => {
        const query = {
          refId: 'A',
          queryType: QueryType.Steps,
          queryByResults: undefined,
          queryBySteps: undefined,
          stepsTake: 1000,
        } as unknown as StepsVariableQuery;
        const result = await datastore.metricFindQuery(query);

        expect(result).toEqual([]);
      });

      it.each([-1, NaN, 10001])('should return empty array if stepsTake value is invalid (%p)', async (invalidStepsTake) => {
        const query = {
          refId: 'A',
          queryType: QueryType.Steps,
          queryByResults: 'PartNumber = "partNumber1"',
          stepsTake: invalidStepsTake,
        } as unknown as StepsVariableQuery;
        const result = await datastore.metricFindQuery(query);

        expect(result).toEqual([]);
      });

      it('should return mapped names when queryByResults is provided and API returns steps', async () => {
        backendServer.fetch
          .mockReturnValue(createFetchResponse({
            steps: [
              { name: 'StepA' },
              { name: 'StepB' }
            ],
            totalCount: 2
          } as QueryStepsResponse));

        const query = { queryByResults: 'PartNumber = "partNumber1"', stepsTake: 1000 } as StepsVariableQuery;
        const result = await datastore.metricFindQuery(query);

        expect(result).toEqual([
          { text: 'StepA', value: 'StepA' },
          { text: 'StepB', value: 'StepB' }
        ]);
      });

      it('should return empty array if API returns no steps', async () => {
        backendServer.fetch
          .mockReturnValue(createFetchResponse({
            steps: [],
            totalCount: 0
          } as QueryStepsResponse));

        const query = { queryByResults: 'PartNumber = "partNumber1"', stepsTake: 1000 } as StepsVariableQuery;
        const result = await datastore.metricFindQuery(query);

        expect(result).toEqual([]);
      });

      it('should return empty array if API throws error', async () => {
        backendServer.fetch.mockImplementationOnce(() => { throw new Error('API error'); });

        const query = { queryByResults: 'PartNumber = "partNumber1"', stepsTake: 1000 } as StepsVariableQuery;
        const result = await datastore.metricFindQuery(query);

        expect(result).toEqual([]);
      });

      it('should use templateSrv.replace for queryByResults and queryBySteps', async () => {
        let resultsQuery = 'PartNumber = "${partNumber}"'
        let stepsQuery = 'stepName = "${step}"'
        templateSrv.replace.mockReturnValueOnce('PartNumber = "partNumber1"').mockReturnValueOnce('stepName = "Step1"');
        backendServer.fetch.mockReturnValue(createFetchResponse({
          steps: [{ name: 'Step1' }],
          totalCount: 1
        } as QueryStepsResponse));

        const query = { queryByResults: resultsQuery, queryBySteps: stepsQuery, stepsTake: 1000 } as StepsVariableQuery;
        await datastore.metricFindQuery(query, { scopedVars: { var: { value: 'replaced' } } } as any);

        expect(templateSrv.replace).toHaveBeenCalledTimes(2);
        expect(templateSrv.replace.mock.calls[0][0]).toBe(resultsQuery);
        expect(templateSrv.replace.mock.calls[1][0]).toBe(stepsQuery);
      });
    });

  const buildQuery = getQueryBuilder<QuerySteps>()({
    refId: 'A',
    queryType: QueryType.Steps,
    outputType: OutputType.Data,
    resultsQuery: 'PartNumber = "partNumber1"'
  });
});
