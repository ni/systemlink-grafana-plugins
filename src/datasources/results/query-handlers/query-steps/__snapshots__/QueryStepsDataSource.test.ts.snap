// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QueryStepsDataSource query should convert properties to Grafana fields 1`] = `
[
  {
    "name": "properties",
    "type": "string",
    "values": [
      "{"key1":"value1","key2":"value2"}",
    ],
  },
]
`;

exports[`QueryStepsDataSource query should convert step measurements to Grafana fields when show measurments is enabled 1`] = `
[
  {
    "name": "Measurement Name",
    "type": "string",
    "values": [
      [
        "Voltage",
        "Voltage",
        "Current",
      ],
    ],
  },
  {
    "name": "Measurement Value",
    "type": "string",
    "values": [
      [
        "3.7",
        "3.7",
        "1.2",
      ],
    ],
  },
  {
    "name": "Status",
    "type": "string",
    "values": [
      [
        "Passed",
        "Passed",
        "Failed",
      ],
    ],
  },
  {
    "name": "Unit",
    "type": "string",
    "values": [
      [
        "V",
        "V",
        "A",
      ],
    ],
  },
  {
    "name": "Low Limit",
    "type": "string",
    "values": [
      [
        "3.5",
        "3.5",
        "1.0",
      ],
    ],
  },
  {
    "name": "High Limit",
    "type": "string",
    "values": [
      [
        "4.0",
        "4.0",
        "1.5",
      ],
    ],
  },
]
`;

exports[`QueryStepsDataSource query should return data for valid data-output-type query 1`] = `
[
  {
    "fields": [
      {
        "name": "name",
        "type": "string",
        "values": [
          "Step 1",
        ],
      },
    ],
    "refId": "A",
  },
]
`;

exports[`QueryStepsDataSource query should return no data when QuerySteps API returns empty array 1`] = `[]`;

exports[`QueryStepsDataSource query should return total count for valid total count output type queries 1`] = `
[
  {
    "fields": [
      {
        "name": "Total count",
        "values": [
          1,
        ],
      },
    ],
    "refId": "A",
  },
]
`;

exports[`QueryStepsDataSource querySteps should return data when there are valid queries 1`] = `
{
  "continuationToken": undefined,
  "steps": [
    {
      "name": "Step 1",
      "properties": {
        "key1": "value1",
        "key2": "value2",
      },
      "stepId": "1",
    },
  ],
  "totalCount": 1,
}
`;
